### PR TITLE
soft fork: implement a way to change the circuit of a chain

### LINKED ISSUE #1196: soft-fork to incorporate the result from the zk-voceremony

### PR DESCRIPTION
**Pull request by altergui**
(see commit messages, please don't squash)

### CODE REVIEW

File: vochain/cometbft.go [hunk 1]
```diff
@@ -288,6 +289,12 @@ func (app *BaseApplication) FinalizeBlock(_ context.Context,
 	if err != nil {
 		return nil, fmt.Errorf("cannot get validators: %w", err)
 	}
+
+	// update ZkCircuit
+	if c := circuit.ZkCircuitForHeight(height + 1); app.TransactionHandler.ZkCircuit != c {
```
Comment (p4u) [inline vochain/cometbft.go]: This should be done on BeginBlock

File: vochain/genesis/genesis.go [hunk 2]
```diff
@@ -14,8 +14,10 @@ var Genesis = map[string]Vochain{
 		SeedNodes: []string{
 			"7440a5b086e16620ce7b13198479016aa2b07988@seed.dev.vocdoni.net:26656",
 		},
-		CircuitsConfigTag: "dev",
-		Genesis:           &devGenesis,
+		Circuits: map[uint32]string{
+			0: "dev",
```
Comment (p4u) [inline vochain/genesis/genesis.go]: I do think tags such as "dev" "prod" etc. Do not work anymore.
I'd just use standard versioning and Git tags.

Comment (altergui) [reply vochain/genesis/genesis.go]: completely agree. funniest of all is that this whole Genesis.Circuits is read nowhere, is completely ignored

File: crypto/zk/circuit/config.go [hunk 3]
```diff
@@ -92,17 +92,16 @@ var CircuitsConfigurations = map[string]*ZkCircuitConfig{
 		WasmHash:                hexToBytes("0x80a73567f6a4655d4332301efcff4bc5711bb48176d1c71fdb1e48df222ac139"),
 		WasmFilename:            "circuit.wasm",
 	},
-	"prod": {
-		URI: "https://raw.githubusercontent.com/vocdoni/" +
-			"zk-franchise-proof-circuit/master",
-		CircuitPath:             "artifacts/zkCensus/dev/160",
+	"v1.0.0": {
+		URI:                     "https://raw.githubusercontent.com/altergui/zk-voceremony",
```
Comment (altergui) [inline crypto/zk/circuit/config.go]: :warning: i published the ceremony result in my fork. don't merge until a formal release is published.

File: vochain/genesis/genesis.go (hunk 2 above)
Comment (altergui) [reply vochain/genesis/genesis.go]: i completely removed the "Tag" concept from circuit package, now everything refers to  "Version"

**Comment by altergui** [comment]
TestVoteCheckZkSNARK is failing, any ideas @lucasmenendez ?

```
2023-11-28T20:32:46.45+01:00 ERR vochain/cometbft.go:210 > checkTx error="voteTx: proof not valid: CSP bundle signature does not match"
2023-11-28T20:32:46.451+01:00 ERR vochain/app.go:234 > rejected tx error="voteTx: proof not valid: CSP bundle signature does not match"
--- FAIL: TestVoteCheckZkSNARK (5.32s)
    transaction_zk_test.go:115: 
        error:
          got non-nil error
        got:
          e"expired sik root provided, generate the proof again"
        stack:
          /home/user/src/vocdoni/vocdoni-node/vochain/transaction_zk_test.go:115
            c.Assert(err, qt.IsNil)
        
```

File: crypto/zk/circuit/config.go (hunk 3 above)
Comment (altergui) [reply crypto/zk/circuit/config.go]: i did publish this in https://github.com/vocdoni/zk-voceremony/pull/139

**Comment by altergui** [comment]
last steps:
* [x] @lucasmenendez merge https://github.com/vocdoni/zk-voceremony/pull/139
* [x] @altergui drop commit https://github.com/vocdoni/vocdoni-node/pull/1205/commits/32eb2feaaf0fb5dec283867a21fc4d1bf0f32a5d
* [x] @altergui bump block numbers in `config/forks.go`
* [x] @altergui mark this PR as ready for review
* [x] review & approve (@lucasmenendez and/or @p4u)
* [x] @altergui merge this PR before those block numbers happen

**Review by p4u — APPROVED** [review]

**Pull request merged**